src/lib/util/graceful-shutdown.ts

```diff
@@ -1,7 +1,60 @@
-import { Logger } from '../logger';
+import { Logger, LogProvider } from '../logger';
 import { IUnleash } from '../types/core';
+import { GenericUnleashError } from '../error/unleash-error';
 
-function registerGracefulShutdown(unleash: IUnleash, logger: Logger): void {
+export interface IGracefulShutdown {
+    registerGracefulShutdownHook(
+        serviceName: string,
+        hook: () => Promise<void>,
+    ): void;
+    resetShutdownHooks(): void;
+    executeShutdownHooks(): Promise<void>;
+}
+
+export class GracefulShutdownHookManager implements IGracefulShutdown {
+    private shutdownHooks: Map<string, () => Promise<void>>;
+    private logger: Logger;
+    constructor(getLogger: LogProvider) {
+        this.shutdownHooks = new Map<string, () => Promise<void>>();
```

Comment:
I think it's worth extracting the shutdown hook type to something like:
```ts
type ShutdownHook = () => void | Promise<void>
```

Reply:
Yeah, I like it

Reply:
handled in b492e8033